File: javascript-project/test/test_MT103_gen.js
const chai = require('chai');
const { expect } = require('chai');
const { DOMParser } = require('@xmldom/xmldom');

// Register the custom assertion
const containXmlTags = require('./xmlTagsAssertion');
chai.use(containXmlTags);

describe('MT103_gen', () => {
  let generateMT103;

  before(() => {
    try {
      ({ generateMT103 } = require('../MT103_gen'));
    } catch (err) {
      generateMT103 = null;
    }
  });

  it('should export a generateMT103 function', () => {
    expect(generateMT103).to.be.a('function');
  });

  it('should generate valid XML', () => {
    const result = generateMT103();
    const parser = new DOMParser();
    const doc = parser.parseFromString(result, 'text/xml');

    // Check for parse errors
    const parseError = doc.getElementsByTagName('parsererror');
    expect(parseError.length).to.equal(0);

    // Check root element exists
    expect(doc.documentElement).to.exist;

  });

  it('XML should be labelled a SwiftPayment record', () => {
    const result = generateMT103();
    const parser = new DOMParser();
    const doc = parser.parseFromString(result, 'text/xml');

    // Check root element name
    expect(doc.documentElement.nodeName).to.equal('SwiftPayment');
  });

  it('should contain all required MT103 XML tags', () => {
    const result = generateMT103();

    const requiredTags = [
      'MessageType',
      'SenderBIC',
      'ReceiverBIC',
      'TransactionReference',
      'ValueDate',
      'Amount',
      'Currency',
      'BeneficiaryName',
      'BeneficiaryAccount',
      'BeneficiaryAddress',
      'RemittanceInformation'
    ];

    // Use the custom assertion
    expect(result).to.containXmlTags(requiredTags);
  });
});

File: python-project/tests/test_MT103_gen.py
import xml.etree.ElementTree as ET
from MT103_gen import generateMT103
from custom_asserts import assert_xml_structure

class TestMT103Gen:
    def test_should_export_a_generate_swift_payment_function(self):
        assert callable(generateMT103)

    def test_should_generate_valid_xml(self):
        result = generateMT103()

        try:
            root = ET.fromstring(result)
            assert root is not None
        except ET.ParseError:
            assert False, "The generated string is not a valid XML document."

    def test_xml_should_be_labelled_a_swift_payment_record(self):
        result = generateMT103()
        try:
            root = ET.fromstring(result)
            assert root.tag == 'SwiftPayment'
        except ET.ParseError:
            assert False, "The valid XML document does not have the expected root tag."

    def test_xml_structure_is_correct(self):
        result = generateMT103()
        expected_tags = ['SwiftPayment', 'MessageType', 'SenderBIC', 'ReceiverBIC', 'TransactionReference', 'ValueDate', 'Amount', 'Currency', 'BeneficiaryName', 'BeneficiaryAccount', 'BeneficiaryAddress', 'RemittanceInformation']
        assert_xml_structure(result, expected_tags)